components/base/kinematicbase/kinematics.go

```diff
@@ -102,56 +107,106 @@ func (ddk *differentialDriveKinematics) GoToInputs(ctx context.Context, desired
 	if err != nil {
 		return err
 	}
-	validRegion, err := ddk.newValidRegionCapsule(current, desired)
+	distErr, headingErr, err := ddk.errorState(current, desired)
+	if err != nil {
+		return err
+	}
+	validRegion, err := ddk.newValidRegionCapsule(current, desired, distErr)
 	if err != nil {
 		return err
 	}
 
+	t := time.NewTimer(timeout)
+	poseChange := make(chan bool)
+	movementErr := make(chan error)
+
 	// this loop polls the error state and issues a corresponding command to move the base to the objective
 	// when the base is within the positional threshold of the goal, exit the loop
-	for err = ctx.Err(); err == nil; err = ctx.Err() {
-		col, err := validRegion.CollidesWith(spatialmath.NewPoint(r3.Vector{X: current[0].Value, Y: current[1].Value}, ""))
-		if err != nil {
-			return err
-		}
-		if !col {
-			return errors.New("base has deviated too far from path")
-		}
+	utils.PanicCapturingGo(func() {
+		prevDistErr := -1
+		prevHeadingErr := math.Inf(-1)
+		var contextErr error
+		for contextErr = ctx.Err(); contextErr == nil; contextErr = ctx.Err() {
+			if poseChanged(prevDistErr, prevHeadingErr, distErr, headingErr) {
+				poseChange <- true
+			}
 
-		// get to the x, y location first - note that from the base's perspective +y is forward
-		desiredHeading := math.Atan2(current[1].Value-desired[1].Value, current[0].Value-desired[0].Value)
-		commanded, err := ddk.issueCommand(ctx, current, []referenceframe.Input{desired[0], desired[1], {desiredHeading}})
-		if err != nil {
-			return err
-		}
+			col, err := validRegion.CollidesWith(spatialmath.NewPoint(r3.Vector{X: current[0].Value, Y: current[1].Value}, ""))
+			if err != nil {
+				movementErr <- err
+				return
+			}
+			if !col {
+				movementErr <- errors.New("base has deviated too far from path")
+				return
+			}
+
+			// get to the x, y location first - note that from the base's perspective +y is forward
+			desiredHeading := math.Atan2(current[1].Value-desired[1].Value, current[0].Value-desired[0].Value)
+			prevDistErr = distErr
+			prevHeadingErr = headingErr
+			distErr, headingErr, err = ddk.errorState(current, []referenceframe.Input{desired[0], desired[1], {desiredHeading}})
```

Comment:
I feel like naming these with "error" is a bit misleading, since any base movement will have a high amount of "error" for any motion, unrelated to how perfectly it executes.

Reply:
the "error" is referring to the error between where the base is and where the goal is. I named it after the fact that `errorState` says it returns distance error and heading error. Not really sure what a better name for this could be

Reply:
I understand that; it's that it's odd to refer to the starting point as an "error state".

Naively I would expect "error" to refer to, say, the deviation from an intended path. Not the state of simply not being at the goal. "Distance" might be a better term here.

Reply:
Disagree Peter, we're writing a control loop here and in control theory is the difference between the desired and the current.

Reply:
This is just a disagreement over what is "desired".

For a control loop that follows a path, I would expect "desired" to be "the optimal path to the goal". As written it's the goal state itself, not a path to it.